src/controller/python/BUILD.gn

```diff
@@ -37,6 +38,12 @@ declare_args() {
   chip_python_supports_stack_locking = chip_system_config_locking != "none"
 }
 
+# Matter Python API supports callbacks registration which then are called
+# from C++ code. This requires the "embed" version of Python library.
```

Comment:
Skimming through [Embedding Python in Another Application](https://docs.python.org/3/extending/embedding.html) doesn't seem to explicitly state that callbacks from a different thread would require the embed version. From what I understand if the main function is in native code this is required, but that is not the case in the CHIP library: The process starts off as a Python program, and the Matter SDK native part gets initialized from it.

Also the [Python API of PyGILState_Ensure](https://docs.python.org/3/c-api/init.html#c.PyGILState_Ensure) doesn't seem to mandate the embed version.

That said, not sure if it is really *not* necessary. But it seems to me? Maybe I am missing something though?

Reply:
> doesn't seem to explicitly state that callbacks from a different thread would require the embed version.

Every time one uses CPython API in the Python extension it is subject of "embedded" Python, which needs "embed" version. The "embed" version is simply a module linked with `libpython.so`.

> PyGILState_Ensure doesn't seem to mandate the embed version.

The `PyGILState_Ensure` is a CPyhton API function, so "embed" is required.

>  From what I understand if the main function is in native code this is required, but that is not the case in the CHIP library: The process starts off as a Python program, and the Matter SDK native part gets initialized from it.

As far as I understand Python, the flow in CHIP Python module looks like this:

1. Python code initializes CHIP Python module (it starts CHIP internal threads)
2. User calls CHIP module function which takes a callback (e.g. a callback called when new CHIP device is found)
3. Registration function returns and Python code is executed on Python native thread
4. CHIP main event loop thread dispatches "new device found event"...

#### Current implementation
5. Python callback is called on CHIP main event loop thread while other Python code runs on Python native thread
6. Callback function modifies some Python object at the same time as the Python main application
7. Python interpreter is in the undefined state...

#### Proper implementation
5. C++ callback calls `PyGILState_Ensure()` to make sure that we can call Python C API
6. C++ callback, calls Python native callback via e.g. `PyObject_CallFunction()`
7. GIL is held by CHIP thread, so no other thread can execute Python code
8. Python callback returns, then C++ callback releases GIL state

Reply:
> Every time one uses CPython API in the Python extension it is subject of "embedded" Python, which needs "embed" version. The "embed" version is simply a module linked with `libpython.so`.

Hm, I see, non-embedding would not link against any Python API, which won't allow us to call any Python functions.
```
$ python3.12-config --ldflags
 -L/usr/lib  -ldl  -lm 
$ python3.12-config --ldflags --embed
 -L/usr/lib -lpython3.12 -ldl  -lm 
 ```

I somehow was under the assumption that we already call into Python from native code. :thinking: 

It seems that we do call `ctypes.pythonapi.Py_IncRef` from Python side, and use `CFUNCTYPE` to generate C ABI Python functions, but nothing really from C++ side indeed, I see.

The [docs of CFUNCTYPE](https://docs.python.org/3.7/library/ctypes.html#ctypes.CFUNCTYPE) says:

> The returned function prototype creates functions that use the standard C calling convention. The function will release the GIL during the call. 

So makes me wonder if there is maybe a way to get GIL handled with ctypes :thinking: 

The [docs about callbacks](https://docs.python.org/3.7/library/ctypes.html#callback-functions) say:

> Also, note that if the callback function is called in a thread created outside of Python’s control (e.g. by the foreign code that calls the callback), ctypes creates a new dummy Python thread on every invocation. This behavior is correct for most purposes, but it means that values stored with [threading.local](https://docs.python.org/3.7/library/threading.html#threading.local) will not survive across different callbacks, even when those calls are made from the same C thread.

So it seems to be safe but not ideal? :thinking: 

I must say, we do have the Python wrapper with the library in use by quite some folks, even with larger setups, and I haven't really observed something which I could pinpoint to Python locking issues. But I guess this doesn't mean anything, as this could be rather subtle.

I did run into these type of issues though (from the callback docs linked above):
> Make sure you keep references to [CFUNCTYPE()](https://docs.python.org/3.7/library/ctypes.html#ctypes.CFUNCTYPE) objects as long as they are used from C code. [ctypes](https://docs.python.org/3.7/library/ctypes.html#module-ctypes) doesn’t, and if you don’t, they may be garbage collected, crashing your program when a callback is made.


Specifically I've fixed such an issue with #30877.

Reply:
I've found this Python discussion quite enlightened: https://bugs.python.org/issue6627.

Essentially, ctypes is taking care of calling `PyGILState_Ensure()`, see: https://github.com/python/cpython/blob/v3.7.17/Modules/_ctypes/callbacks.c#L140.

Reply:
> Also, note that if the callback function is called in a thread created outside of Python’s control (e.g. by the foreign code that calls the callback), ctypes creates a new dummy Python thread on every invocation.

Hmm... interesting.... I've been doing some tests (it was like half a year ago) to check whether calling callbacks without `PyGILState_Ensure()` is OK. And if I remember correctly I ended up with two Linux threads running Python code at the same time. But maybe I did something wrong :/

Anyway, using `ctypes.pythonapi.Py_IncRef` instead of clean call from C code adds a lot of maintenance burden IMO. It would be much better to have a clean Python API exposed by C module, i.e. you can pass Python object (implementing some interface) as a closure to C API without playing around with function callbacks wrappers. A lot of could could be removed.

However, when speaking of Python bindings there are better approaches. One of these is https://github.com/pybind/pybind11

Reply:
> And if I remember correctly I ended up with two Linux threads running Python code at the same time. But maybe I did something wrong :/

There is definitely quite a bit of (seemingly) concurrent Python code execution going on, e.g. we print the thread in our Python Matter server (which uses the current wheels) in logs:

```
2024-05-16 11:48:21.218 (MainThread) INFO [matter_server.server.device_controller.node_21] Setting up attributes and events subscription.
2024-05-16 11:48:21.218 (Dummy-2) CHIP_PROGRESS [chip.native.DIS] Found an existing secure session to [1:0000000000000015]!
```

All logging of the SDK goes through Python (since logs get redirected through Python logging), so each log message is from a different Thread.

But even these callbacks are [LogRedirectCallback_t](https://github.com/project-chip/connectedhomeip/blob/v1.3.0.0/src/controller/python/chip/logging/types.py#L20) (`CFUNCTYPE`) [callbacks](https://github.com/project-chip/connectedhomeip/blob/v1.3.0.0/src/controller/python/chip/logging/__init__.py#L29-L30), so they should take care of the GIL as well. So things really shouldn't run truly concurrent.

> Anyway, using ctypes.pythonapi.Py_IncRef instead of clean call from C code adds a lot of maintenance burden IMO. It would be much better to have a clean Python API exposed by C module, i.e. you can pass Python object (implementing some interface) as a closure to C API without playing around with function callbacks wrappers. A lot of code could be removed :)
>
> However, when speaking of Python bindings there are better approaches. One of these is https://github.com/pybind/pybind11

I agree, the current ctypes based bindings are not ideal.

Whenever mixing native code with Python (which uses a garbage collector), a certain amount of manual memory management will be necessary most likely. 

But maybe pybind11 can use C++ smart pointers or similar C++ features to make this (more) transparently? :thinking: